---
title: "LFMM"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r, PCA on environmental variables}

outlierdf
fulldf <- fulldata[,8:26]
env_outdf <- as.matrix(fulldf[-c(27,28,31,32,61,64,71,72,73,74,75,76,77), ])
colnames(env_outdf) <- c("Annual Mean Temperauture", "Mean Tempereature of Warmest Month", "Mean Temperature of Coldest Quarter", "Annual Precipitation", "Precipitation of Wettest Month", "Precipitation of Driest Month", "Precipitation Seasonality", "Precipitation of Wettest Quarter", "Precipitation of Driest Quarter", "Precipitation of Warmest Quarter", "Precipitation of Coldest Quarter", "Mean Diurnal Range", "Isothermality", "Temperature Seasonality", "Max Temperature of Warmest Month", "Min Temperature of Coldest Month", "Temperature Annual Range", "Mean Temperature of Wettest Quarter", "Mean Temperature of Coldest Quarter")

env_outdf1 <- as.data.frame(env_outdf) %>% 
  relocate("Temperature Seasonality", "Max Temperature of Warmest Month", "Min Temperature of Coldest Month", "Temperature Annual Range", "Mean Temperature of Wettest Quarter", "Mean Temperature of Coldest Quarter")
env_outdf2 <- as.matrix(env_outdf1)
scaled_envdf <- scale(env_outdf)
env_pca <- prcomp(scaled_envdf, scale = TRUE)
summary(env_pca)
pc_loadings <- env_pca$rotation[,1:3]

# view the loadings as a heatmap
heatmap(pc_loadings, Rowv=NA, Colv=NA, scale="none")
envPC1 <- predict(env_pca)[, 1]
envPC2 <- predict(env_pca)[, 2]

```

```{r, PCA of environment data}
env_pca

env_out1 <- data.frame(env_pca$x)
env_out1$plotx <- env_out1[,1]
env_out1$ploty <- env_out1[,2]
#Calculate PC scores
env_eigenvalues <- env_pca$sdev^2
env_prop_eigen <- env_eigenvalues/sum(env_eigenvalues)

Latitude = outlierdf$Latitude
Host = outlierdf$Host
#https://sjmgarnier.github.io/viridis/reference/scale_viridis.html
env_PCA <- ggplot(env_out1, aes(x=plotx, y = ploty, color=Latitude)) + geom_point(size = 3.0, aes(shape = Host)) + xlab("PC1 (56.5%)") + ylab("PC2 (31.2%)") + scale_colour_gradientn(colours = viridis(100)) + theme_bw() + theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank())
env_PCA
```

```{r, LFMM}
#https://cran.r-project.org/web/packages/lfmm/vignettes/lfmm.html
#install.packages('lfmm')

library(lfmm)
?lfmm
outlierdf

pc <- prcomp(num_out_df)
plot(pc$sdev[1:20]^2, xlab = 'PC', ylab = "Variance explained")
points(3,pc$sdev[6]^2, type = "h", lwd = 3, col = "blue")

mod.lfmm <- lfmm_ridge(Y = num_out_df,
                       X = envPC1 + envPC2,
                       K = 3)
pv <- lfmm_test(Y = num_out_df, 
                 X = envPC1 + envPC2, 
                 lfmm = mod.lfmm, 
                 calibrate = "gif")
pv <- as.data.frame(pv)

pvalues <- pv$calibrated.pvalue 


#Compute direct effect sizes using the lfmm_ridge function
mod <- lfmm_ridge(Y = num_out_df,
                  X = envPC1 + envPC2,
                  K = 3)

pred <- predict_lfmm(Y = num_out_df,
                     X = envPC1 + envPC2,
                     fdr.level = 0.1,
                     mod)
#Adjust p-value with fdr correction
p.adj <- p.adjust(pvalues, method="fdr")
alpha <- 0.05
out.lfmm <- which(p.adj < alpha)
length(out.lfmm)


sampleID <- as.data.frame(rownames(pv))
sampleID$rows <- seq(nrow(sampleID))
p.out.lfmm <- sampleID[sampleID$rows %in% out.lfmm, ]
p.GLM <- sampleID[sampleID$rows %in% label_snps, ]
#Dataframe with outlier loci 
sample.out.lfmm <- as.data.frame(p.out.lfmm[, 1])


write.table(sample.out.lfmm, file = "LFMM.outlier.loci.csv", sep = ",", col.names = FALSE, row.names = FALSE)
qqplot(rexp(length(pvalues), rate = log(10)),
       -log10(pvalues), xlab = "Expected quantile",
       pch = 19, cex = .4)
abline(0,1)


threshold <- 0.005  # set the p-value threshold for labeling SNPs

#Manhattan Plot 
plot(-log10(p.adj), 
      pch = 19, 
      cex = .6, 
      xlab = "snp position", ylab = "-log(pvalue)",
      col = "grey")
 points(out.lfmm, #highlight snps of interest
        -log10(p.adj)[out.lfmm], 
        type = "p", 
        col = "blue")
 label_snps <- which(p.adj < threshold)
 text(out.lfmm[label_snps], -log10(pvalues)[out.lfmm][label_snps], labels = names(pvalues)[label_snps], pos = 2)


hostdf <- outlierdf[,4:6]
hostdf <- hostdf[-c(2)]
hostdf %>% group_by(Host) %>% summarise(mean.lat = mean(Latitude), median.latitude = median(Latitude)) 
t.test(((hostdf %>% filter(Host == 'Bishop'))$Latitude), ((hostdf %>% filter(Host == 'Monterey'))$Latitude))








```

```{r, PCA with 120 outlier loci}

columns_to_select <- p.out.lfmm$`rownames(pv)`
out_subset_df <- num_out_df[, columns_to_select]

scaled_out_subset_df <- scale(out_subset_df)
subset_pca <- prcomp(scaled_out_subset_df, scale = TRUE)
summary(subset_pca)


subset_out <- data.frame(subset_pca$x)
subset_out$plotx <- subset_out[,1]
subset_out$ploty <- subset_out[,2]
#Calculate PC scores
subset_eigenvalues <- subset_pca$sdev^2
env_prop_eigen <- subset_eigenvalues/sum(subset_eigenvalues)

Latitude = outlierdf$Latitude
Host = outlierdf$Host
#https://sjmgarnier.github.io/viridis/reference/scale_viridis.html
subset_PCA <- ggplot(subset_out, aes(x=plotx, y = ploty, color=Latitude)) + geom_point(size = 3.0, aes(shape = Host)) + xlab("PC1 (42.1%)") + ylab("PC2 (10.8%)") + scale_colour_gradientn(colours = viridis(100)) + theme_bw() + theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank())
subset_PCA
```


```{r, candidate genes}
Importance of components:
                          PC1    PC2     PC3     PC4     PC5     PC6     PC7     PC8     PC9
Standard deviation     10.333 5.5059 5.12779 3.29726 2.74516 2.51364 2.38571 2.23422 2.21226
Proportion of Variance  0.372 0.1056 0.09162 0.03788 0.02626 0.02202 0.01983 0.01739 0.01705
Cumulative Proportion   0.372 0.4777 0.56928 0.60716 0.63342 0.65543 0.67526 0.69266 0.70971
                          PC10    PC11    PC12    PC13    PC14    PC15    PC16    PC17    PC18
Standard deviation     2.11259 1.96242 1.88175 1.84961 1.81882 1.74326 1.73684 1.72457 1.68762
Proportion of Variance 0.01555 0.01342 0.01234 0.01192 0.01153 0.01059 0.01051 0.01036 0.00992
Cumulative Proportion  0.72526 0.73868 0.75102 0.76294 0.77446 0.78505 0.79556 0.80592 0.81585
                          PC19   PC20    PC21    PC22    PC23    PC24    PC25    PC26    PC27
Standard deviation     1.64325 1.5890 1.57809 1.52242 1.49464 1.46190 1.45843 1.38574 1.35312
Proportion of Variance 0.00941 0.0088 0.00868 0.00808 0.00778 0.00745 0.00741 0.00669 0.00638
Cumulative Proportion  0.82526 0.8340 0.84273 0.85081 0.85859 0.86604 0.87345 0.88014 0.88652
                          PC28    PC29    PC30    PC31   PC32    PC33    PC34    PC35    PC36
Standard deviation     1.32645 1.28659 1.27157 1.25749 1.2334 1.18941 1.17294 1.13877 1.12269
Proportion of Variance 0.00613 0.00577 0.00563 0.00551 0.0053 0.00493 0.00479 0.00452 0.00439
Cumulative Proportion  0.89265 0.89842 0.90405 0.90956 0.9149 0.91979 0.92458 0.92910 0.93349
                          PC37    PC38   PC39   PC40    PC41   PC42    PC43    PC44    PC45
Standard deviation     1.08968 1.07284 1.0445 1.0171 0.99963 0.9880 0.94618 0.92877 0.90847
Proportion of Variance 0.00414 0.00401 0.0038 0.0036 0.00348 0.0034 0.00312 0.00301 0.00288
Cumulative Proportion  0.93763 0.94164 0.9454 0.9490 0.95253 0.9559 0.95905 0.96206 0.96493
                          PC46    PC47    PC48    PC49    PC50    PC51    PC52    PC53    PC54
Standard deviation     0.87452 0.87297 0.85285 0.82677 0.81400 0.79341 0.77888 0.72849 0.72278
Proportion of Variance 0.00266 0.00266 0.00253 0.00238 0.00231 0.00219 0.00211 0.00185 0.00182
Cumulative Proportion  0.96760 0.97025 0.97279 0.97517 0.97748 0.97967 0.98178 0.98363 0.98545
                         PC55    PC56    PC57    PC58    PC59   PC60    PC61    PC62   PC63
Standard deviation     0.6991 0.67596 0.65369 0.63732 0.62265 0.5859 0.56373 0.52440 0.5071
Proportion of Variance 0.0017 0.00159 0.00149 0.00142 0.00135 0.0012 0.00111 0.00096 0.0009
Cumulative Proportion  0.9872 0.98875 0.99024 0.99165 0.99300 0.9942 0.99531 0.99626 0.9972
                          PC64    PC65    PC66    PC67    PC68      PC69
Standard deviation     0.47532 0.44646 0.44476 0.39080 0.19815 4.589e-15
Proportion of Variance 0.00079 0.00069 0.00069 0.00053 0.00014 0.000e+00
Cumulative Proportion  0.99795 0.99864 0.99933 0.99986 1.00000 1.000e+00
R Console


genes <- as.data.frame(read.delim("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/Suipun1/Suipun1_GeneCatalog_genes_20200424.gff", header=F, comment.char="#"))

candidate_snps <- as.data.frame(read.csv("./LFMM.outlier.loci.csv", sep = ",", header = FALSE))
colnames(candidate_snps)[1] <- "scaffold"
colnames(candidate_snps)[2] <- "pos"
write.table(candidate_snps, file = "./candidate_snps_lfmm.csv", sep = ",", col.names = TRUE)

genes$V1 <-as.character(genes$V1)
candidate_snps$scaffold <- as.character(candidate_snps$scaffold)
scaffold_matches <- match(genes$V1, candidate_snps$scaffold)
length(unique(scaffold_matches))

candidate_genes <- genes[!is.na(scaffold_matches) & 
                         genes$V4 <= 50 + candidate_snps$pos[scaffold_matches] & 
                         genes$V5 >= 50 + candidate_snps$pos[scaffold_matches], ]

gff_scaffolds <- unique(genes$V1)
missing_scaffolds <- candidate_snps$scaffold[!(candidate_snps$scaffold %in% gff_scaffolds)]

if (length(missing_scaffolds) > 0) {
  message("The following scaffolds are not present in the GFF file: ", paste(missing_scaffolds, collapse = ", "))
} else {
  message("All scaffolds in the candidate SNP list are present in the GFF file.")
}

```

```{r, KOG assingments}

#removing some columns and seperating INFO column
KOG.df <- as.data.frame(read.table("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/Suipun1/Suipun1_GeneCatalog_proteins_20200424_KOG.tab.gz", header=F, sep="\t", fill = TRUE, quote = ""))


candidate_genes2 <- data.frame(candidate_genes$V1, candidate_genes$V4, candidate_genes$V5, candidate_genes$V3, do.call(rbind, strsplit(candidate_genes$V9, split = ";", fixed = TRUE)))
colnames(candidate_genes2)[1] = "Chrom"
colnames(candidate_genes2)[2] = "Start_Pos"
colnames(candidate_genes2)[3] = "End_pos"
colnames(candidate_genes2)[4] = "Type"
colnames(candidate_genes2)[5] = "Name"
colnames(candidate_genes2)[6] = "Protein_ID"
colnames(candidate_genes2)[7] = "Product_Name"
candidate_genes2$Protein_ID <- gsub("proteinId", "", candidate_genes2$Protein_ID)

#candidate_genes3 is the cleaned df
candidate_genes3 <- candidate_genes2[!(candidate_genes2$Type=="exon"), ]
candidate_genes3$Protein_ID <- as.numeric(candidate_genes3$Protein_ID)


################################################
colnames(KOG.df)[1] = "TransciptID"
colnames(KOG.df)[2] = "ProteinID"
colnames(KOG.df)[3] = "KOGterm"
colnames(KOG.df)[4] = "KOGtermDescription"
colnames(KOG.df)[5] = "Type"
colnames(KOG.df)[6] = "Process"
################################################

#Pull goterms that are in gene_list2 dataframe
KOG_list <- as.data.frame(KOG.df %>% 
  filter(KOG.df$ProteinID %in% candidate_genes3$Protein_ID))
GO_list <- GO.df %>% 
  filter(GO.df$proteinId %in% candidate_genes3$Protein_ID)

length(unique(GO_list$proteinId))

write.table(KOG_list, file = "./KOG_outlier_snps.csv", sep = ",", col.names = TRUE)
```

```{r, Outflank, LFMM, PCAdapat}
outflank_snp_pos <- as.numeric(outflank_snp_pos)
AdaptFlank_outliers <- outliers_pcadapt[outliers_pcadapt %in% outflank_snp_pos]

LFMM_out_total <- out.lfmm[out.lfmm %in% AdaptFlank_outliers]


sampleID <- as.data.frame(rownames(pv))
sampleID$rows <- seq(nrow(sampleID))
p.out.3way <- sampleID[sampleID$rows %in% LFMM_out_total, ]

#Dataframe with outlier loci 
sample.out.3way <- as.data.frame(p.out.3way[1])
colnames(sample.out.3way)[1] <- "name"
sample.out.3way$position <- gsub(".*_(\\d+)$", "\\1", sample.out.3way$name)
sample.out.3way$name <- gsub("_\\d+$", "", sample.out.3way$name)

#dataframe with scaffold column and pos column
sample.out.3way
colnames(sample.out.3way)[1] <- "scaffold"
colnames(sample.out.3way)[2] <- "pos"

genes$V1 <-as.character(genes$V1)
sample.out.3way$scaffold <- as.character(sample.out.3way$scaffold)
scaffold_matches3way <- match(genes$V1, sample.out.3way$scaffold)
length(unique(scaffold_matches3way))

candidate_genes3way <- genes[!is.na(scaffold_matches3way) & 
                         genes$V4 <= sample.out.3way$pos[scaffold_matches3way] & 
                         genes$V5 >= sample.out.3way$pos[scaffold_matches3way], ]



cand_genes <- data.frame(candidate_genes3way$V1, candidate_genes3way$V4, candidate_genes3way$V5, candidate_genes3way$V3, do.call(rbind, strsplit(candidate_genes3way$V9, split = ";", fixed = TRUE)))
colnames(cand_genes)[1] = "Chrom"
colnames(cand_genes)[2] = "Start_Pos"
colnames(cand_genes)[3] = "End_pos"
colnames(cand_genes)[4] = "Type"
colnames(cand_genes)[5] = "Name"
colnames(cand_genes)[6] = "Protein_ID"
colnames(cand_genes)[7] = "Product_Name"
cand_genes$Protein_ID <- gsub("proteinId", "", cand_genes$Protein_ID)

#candidate_genes3 is the cleaned df
cand_genes2 <- cand_genes[!(cand_genes$Type=="exon"), ]
cand_genes2$Protein_ID <- as.numeric(cand_genes2$Protein_ID)


################################################
colnames(KOG.df)[1] = "TransciptID"
colnames(KOG.df)[2] = "ProteinID"
colnames(KOG.df)[3] = "KOGterm"
colnames(KOG.df)[4] = "KOGtermDescription"
colnames(KOG.df)[5] = "Type"
colnames(KOG.df)[6] = "Process"
################################################

#Pull goterms that are in gene_list2 dataframe
KOG_list3way <- as.data.frame(KOG.df %>% 
  filter(KOG.df$ProteinID %in% cand_genes2$Protein_ID))
GO_list3way <- GO.df %>% 
  filter(GO.df$proteinId %in% cand_genes2$Protein_ID)

length(unique(KOG_list$ProteinID))




```

```{r, GLM on candidate snps}
cols_of_interest <- p.GLM$`rownames(pv)`


df_subset <- num_out_df[, cols_of_interest]
df_full <- cbind(env_data, df_subset)

snp_subset <- df_subset[,4]
glm1 <- glm(snp_subset ~ wc2.1_2.5m_bio_1, data = df_full, family = poisson)
#summary(glm1)

plot_data <- data.frame(snp = snp_subset, phenotype = df_full$wc2.1_2.5m_bio_1)

ggplot(plot_data, aes(x = phenotype, y = snp)) +
  geom_jitter(width = 0.4, height = 0.2, alpha = 0.5) +
  geom_point() +
  geom_smooth(method = "glm", se = TRUE) +
  xlab("Mean Annual Temperature") +
  ylab("Allele Frequency")


df1 <- data.frame(
  temp = c(10,20,25,30,40),
  biomass = c(5, 14,21,8,2)
)

ggplot(df1, aes(x=temp, y=biomass)) + 
  geom_point(size=3) + 
  geom_line(size =1) + 
  labs(x="Temperature (C)", y="Biomass (mg)")

```

```{r, Kinship Matrix}
library(GenomicRanges)
library(SNPRelate)

outlierdf
gene_df <- candidate_genes3[ ,1:3]
snp_df <- num_out_df
rownames(snp_df) <- outlierdf[,1]

rownames(out_subset_df) <- outlierdf[,2]
out_matrix <- as.matrix(out_subset_df)
heatmap(out_matrix, scale = "none")

snp_df <- num_out_df[, grep("scaffold_109|scaffold_13|scaffold_134|scaffold_212|scaffold_286|scaffold_363|scaffold_383|scaffold_387|scaffold_401|scaffold_47|scaffold_492|scaffold_62|scaffold_621scaffold_70", names(num_out_df))]
rownames(snp_df) <- outlierdf[,2]
snp_df1 <- rownames_to_column(snp_df, var = "sample_id")
df_snps_long <- pivot_longer(snp_df1, cols = starts_with("scaffold"), 
                             names_to = "snp_id", values_to = "genotype")
df_snps <- separate(df_snps_long, col = snp_id, into = c("scaffold", "number", "position"), sep = "_")
df_snps$scaffold <- paste(df_snps$scaffold, df_snps$number, sep = "_")
df_snps <- df_snps[,-c(3)]
df_snps$position <- as.numeric(df_snps$position)

gene_df
colnames(gene_df)[1] <- "scaffold"
colnames(gene_df)[2] <- "start_position"
colnames(gene_df)[3] <- "end_position"

snp_granges <- with(df_snps, GRanges(seqnames = scaffold,
                                    ranges = IRanges(start = position,
                                    end = position)))
gene_granges <- with(gene_df, GRanges(seqnames = scaffold,
                                      ranges = IRanges(start = start_position,
                                                       end = end_position)))
overlap_counts <- countOverlaps(snp_granges, gene_granges)

df_snps$gene_counts <- overlap_counts
filtered_snps_df <- subset(df_snps, gene_counts > 0)

summary_df <- filtered_snps_df %>%
  group_by(sample_id, scaffold) %>%
  summarize(num_snps = sum(gene_counts), .groups = "drop")

#Generate HeatMap
heatmap_df <- pivot_wider(summary_df, names_from = scaffold, values_from = num_snps, values_fill = 0)
rownames(heatmap_df) <- NULL

heatmap_data <- data.matrix(heatmap_df)
heatmap(heatmap_data, Rowv = NA, Colv = NA, col = colorRampPalette(c("white", "blue"))(100), scale = "none")

install.packages('rrBLUP')
library(rrBLUP)
out_matrix

kinship_matrix <- A.mat(out_matrix)
heatmap(kinship_matrix, col = colorRampPalette(c("white", "blue"))(100),
        cexRow = 0.4, cexCol = 0.2)

##############################################################
sampleID <- as.data.frame(rownames(pv))
sampleID$rows <- seq(nrow(sampleID))
p.out.lfmm1 <- sampleID[sampleID$rows %in% out.lfmm1, ]

#Dataframe with outlier loci 
sample.out.lfmm1 <- as.data.frame(p.out.lfmm1[, 1])

columns_to_select1 <- p.out.lfmm1$`rownames(pv)`
out_subset_df1 <- num_out_df[, columns_to_select1]
out_matrix1 <- as.matrix(out_subset_df1)
kinship_matrix1 <- A.mat(out_matrix1)
heatmap(kinship_matrix1, col = colorRampPalette(c("white", "blue"))(100),
        cexRow = 0.4, cexCol = 0.4)
as
```

