---
title: "R Notebook"
output: html_notebook
---

Github token: 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

```{r}
library(dplyr)
library(tidyr)
library(data.table)
library(stats)
library(viridis)
library(zoo)
library(tidyverse)
library(raster)
library(rgdal)
library(ggpubr)
library(broom)
library(AICcmodavg)
library(rstatix)
library(vegan)
library(factoextra)
library(qqman) 
library(SNPfiltR)
library(LDlinkR)


```

```{r, read files}
table <- fread(file = "/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/SP.merged.filtered.table.012.tsv", sep = '\t', header = FALSE)
indv <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/sampleID.csv")
pos <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/SP.merged.filtered.table.012.pos.tsv", sep = '\t', header = FALSE)
coordinates <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/SNPlocation.csv")
Bio.clim.t.mean.files <- list.files("/Users/jyeam/Documents/BioClim_variables_2.5m_bio", ".tif", full.names=TRUE)
host <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/HostType.csv")
specimen <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/SpecimenID.csv")
population <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/Pop_Gen_DATA.csv")
```

```{r, data formatting}
pos$V1 <- paste(pos$V1, pos$V2, sep = "_")
pos <- subset(pos, select = -c(V2))

geno <- subset(table, select = -c(V1))
colnames(geno) <- pos$V1

geno.frame <- cbind(indv, geno) %>% 
  as.data.frame()

coordinates_df <- cbind(coordinates, geno.frame)
coordinates_df <- subset(coordinates_df, select = -c(Sample))

##stack them together, this is the critical step
BIOCLIM.stack <- stack(Bio.clim.t.mean.files)
library(dplyr)

coord_site <- coordinates_df %>% 
  dplyr::select(Sample,Longitude,Latitude) %>% 
  na.omit() %>% 
  column_to_rownames(var='Sample') %>% 
  data.frame
# extract values for given Long/lat as specified for each coordinate
BIOCLIM_ext <- raster::extract(BIOCLIM.stack, coord_site, df=T)
BIOCLIM_character <- BIOCLIM_ext %>% 
  mutate(`ID`=as.character(`ID`))
##now add the identifying domainIDback on to this: 
clim_data_NEON<- cbind(coord_site,BIOCLIM_character) %>% 
  dplyr::select(-ID) %>% 
  rownames_to_column(var="Sample")

coordinates_df[1:3,541745:541752]
##confirm that the rows are retained as desired: 
full <- cbind(coordinates_df,clim_data_NEON) %>% #combine datasets
  relocate(wc2.1_2.5m_bio_1:wc2.1_2.5m_bio_9, .before = scaffold_1_6796) #moving bioclim data to front
  # this was removed from full command %>% dplyr::select(-c(Sample.1, Longitude.1, Latitude.1))
  
fulldata1 <- cbind(host,specimen, population$Population, full)
colnames(fulldata1)[5] <- "Population"
fulldata2 <- fulldata1[,1:541776] 
fulldata <- dplyr::select(fulldata2,-Sample.ID) %>% 
  relocate(Sample, SpecimenID)


write.table(clim_data_NEON,file="clim_data_NEON.csv",sep=",")


numerical <- fulldata[,27:ncol(fulldata)]
numdf <- numerical[ , which(apply(numerical, 2, var) != 0)] #remove zero variance columns from dataset

```

```{r, PCA}


pca_result <- prcomp(numdf, scale = TRUE)
summary(pca_result)
#http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/118-principal-component-analysis-in-r-prcomp-vs-princomp/
PCA <- fviz_pca_ind(pca_result,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

df_out <- data.frame(pca_result$x)
df_out$plotx <- df_out[,1]
df_out$ploty <- df_out[,2]

# I don't know what this is for (below)
#latitude <- as.character(coord_df$Latitude)


#https://sjmgarnier.github.io/viridis/reference/scale_viridis.html
p <- ggplot(df_out, aes(x=plotx, y = ploty, color=fulldata$wc2.1_2.5m_bio_12)) + geom_point() + xlab("PC1") + ylab("PC2") + scale_colour_gradientn(colours = magma(100)) + theme_bw() + theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank())

#Remove outliers
masterdata <- fulldata[-c(61,64,73,74,75,76,77), ] %>% #samples had low read count so discard
  relocate(Sample)
masternum <- masterdata[,25:ncol(masterdata)]
masternumerical <- masternum[ , which(apply(masternum, 2, var) != 0)]
pca_master <- prcomp(masternumerical, scale = TRUE)

PCAmaster <- fviz_pca_ind(pca_master,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

fulldata[19, 4] = "Monterey"
outlierdf <- fulldata[-c(27,28,31,32,61,64,71,72,73,74,75,76,77), ]
out_df <- outlierdf %>% 
  relocate(Sample) 

rownames(out_df) <- NULL 
numerical_out <- out_df[,27:ncol(out_df)]
num_out_df <- numerical_out[ , which(apply(numerical_out, 2, var) != 0)] #remove zero variance columns from dataset
save(num_out_df, file = "num_out_df.Rda")
pca_out <- prcomp(num_out_df, scale = TRUE)

PCA1 <- fviz_pca_ind(pca_out,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )
df_out1 <- data.frame(pca_out$x)
df_out1$plotx <- df_out1[,1]
df_out1$ploty <- df_out1[,2]
#Calculate PC scores
eigenvalues <- pca_out$sdev^2
prop_eigen <- eigenvalues/sum(eigenvalues)

Latitude = outlierdf$Latitude
Host = outlierdf$Host
#https://sjmgarnier.github.io/viridis/reference/scale_viridis.html
p1 <- ggplot(df_out1, aes(x=plotx, y = ploty, color=Latitude)) + geom_point(size = 3.0, aes(shape = Host)) + xlab("PC1 (3.14%)") + ylab("PC2 (2.82%)") + scale_colour_gradientn(colours = viridis(100)) + theme_bw() + theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank())
p1

```

```{r, ADONIS}
#Analysis of variance using distance matrices
num_out_df
MAT = as.numeric(outlierdf$wc2.1_2.5m_bio_1)
ISO = as.numeric(outlierdf$wc2.1_2.5m_bio_3)
SEASONAL_TEMP = as.numeric(outlierdf$wc2.1_2.5m_bio_4)
TEMP_RANGE = as.numeric(outlierdf$wc2.1_2.5m_bio_7)
PRECIP = as.numeric(outlierdf$wc2.1_2.5m_bio_12)
SEASONAL_PRECIP = as.numeric(outlierdf$wc2.1_2.5m_bio_15)
Host <- as.factor(outlierdf$Host)

outlierdf$Bishop <- ifelse(outlierdf$Host == 'Bishop', 1, 0)
outlierdf$Monterey <- ifelse(outlierdf$Host == 'Monterey', 1, 0)
outlierdf <- outlierdf %>% 
  relocate(Bishop, Monterey, .after = Host) %>% 
  relocate(Sample)

#adonis <- adonis(num_out_df ~ MAT + ISO + SEASONAL_TEMP + TEMP_RANGE + PRECIP + SEASONAL_PRECIP + Host, outlierdf, permutations = 999, method = "bray", strata = NULL)

#adonis1 <- adonis(num_out_df ~ MAT + ISO + PRECIP + SEASONAL_PRECIP, outlierdf, permutations = 999, method = "bray", strata = NULL)

adonis2 <- adonis(num_out_df ~ MAT + ISO + PRECIP + SEASONAL_PRECIP + Host, outlierdf, permutations = 999, method = "bray", strata = NULL)


library(gdm)
citation("gdm")

```

```{r, Dendrogram}
#http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning
# other that I didn't really use: https://www.rpubs.com/dvallslanaquera/clustering
install.packages('ggdendro')
install.packages('ape')
if(!require(dendextend)) install.packages("dendextend")
library("dendextend")
?dendextend
library(ggdendro)
library(cluster)
library(factoextra)

dendrodata <- outlierdf[,-1]
rownames(dendrodata) <- dendrodata[,1]
dendrodf <- as.data.frame(dendrodata[,27:ncol(dendrodata)])
dendrodf[ , 541740:541747]
dd <- dist(scale(dendrodf), method = "euclidean")
hc <- hclust(dd, method = "ward.D2")

###############
fulldata
fulldata_num <- fulldata[,26:ncol(fulldata)]
#https://stackoverflow.com/questions/27485549/how-to-colour-the-labels-of-a-dendrogram-by-an-additional-factor-variable-in-r
hcd <- as.dendrogram(hc)
labels_colors(hcd)
par(mfrow = c(1,2))
plot(hcd, main = "Original dend")
#adding some color by tree association
colors_to_use <- unclass(as.factor(dendrodata[,3]))
colors_to_use <- colors_to_use[order.dendrogram(hcd)]
labels_colors(hcd) <- colors_to_use
plot(hcd, main = "Cluster Dendrogram Grouped by Host Association")
###############


plot(hclust(dd, method = "ward.D2"))
rect.hclust(hc, k=4, border = 2:5)


ggdendrogram(hc, rotate = FALSE, size = 2)
ddata <- dendro_data(hcd, type = "rectangle")
dendro_plot <- ggplot(segment(ddata)) +
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) + 
  coord_flip() + 
  scale_y_reverse(expand = c(0.2, 0))

plot(hc)
rect.hclust(hc, k = 4, border = 2:5)
hcd <- as.dendrogram(hc)
plot(hc)
plot(hcd)
png("ClusterDendrogram.png",width=13,height=5,units="in",res=4000)
plot(hc)
dev.off()
dendrogram <- plot(hcd)

  

labels(hcd) <- masterdata$SpecimenID
labels(hcd)
plot(hc, labels = NULL, cex=0.8, hang = 0.1, 
     main = "Cluster dendrogram", sub = NULL,
     xlab = NULL, ylab = "Height")

#Divide into Subgroups
sub_grp <- cutree(hc, k =4)
table(sub_grp)

outlierdf[19, 4] = "Monterey"

plot(as.phylo(hc), cex = 0.5, label.offset = 0.5)

plot(as.phylo(hc), type = "unrooted", cex = 0.6,
     no.margin = TRUE)

nodePar <- list(lab.cex = 0.6, pch = c(NA, 19), cex = 0.7, col = "blue")


```

```{r, GDM}

library(gdm)

######################## 
#Create bishop df
Bishop <- filter(outlierdf, Host =="Bishop")
Sample <- Bishop[,1]
Bishopdf <- Bishop[,25:ncol(Bishop)]
Bishopdf <- Bishopdf[ , which(apply(Bishopdf, 2, var) != 0)]
bishopdf <- cbind(Sample, Bishopdf)

bishopenv <- subset(Bishop, select = c(Sample, Latitude, Longitude, wc2.1_2.5m_bio_1, wc2.1_2.5m_bio_12, wc2.1_2.5m_bio_4, wc2.1_2.5m_bio_3, wc2.1_2.5m_bio_15))
Bishop_sub <- df.sample
Bishop_sub <- bishopdf[,1:50000]

#Create Monterey df
Monterey <- filter(outlierdf, Host == "Monterey")
Sample <- Monterey[,1]
Montereydf <- Monterey[,25:ncol(Monterey)]
Montereydf <- Montereydf[ , which(apply(Bishopdf, 2, var) != 0)]
montereydf <- cbind(Sample, Montereydf)
montereyenv <- subset(Monterey, select = c(Sample, Latitude, Longitude, wc2.1_2.5m_bio_1, wc2.1_2.5m_bio_12, wc2.1_2.5m_bio_4, wc2.1_2.5m_bio_3, wc2.1_2.5m_bio_15))
Monterey_sub <- montereydf[,1:50000]

GDM_bish <- formatsitepair(bioData = Bishop_sub, bioFormat = 1, dist="bray", abundance = T, siteColumn = "Sample", XColumn= "Latitude", YColumn = "Longitude", predData = bishopenv)
GDM_mont <- formatsitepair(bioData = Monterey_sub, bioFormat = 1, dist="bray", abundance = T, siteColumn = "Sample", XColumn= "Latitude", YColumn = "Longitude", predData = montereyenv)
GDM_bish <- formatsitepair(bioData = Bishop_sub, bioFormat = 1, dist="bray", abundance = T, siteColumn = "Sample", XColumn= "Latitude", YColumn = "Longitude", predData = bishopenv)
############################


#BIO1 = Annual Mean Temperature
#BIO3 = Isothermality (BIO2/BIO7) (×100)
#BIO4 = Temperature Seasonality (standard deviation ×100)
#BIO5 = Max Temperature of Warmest Month
#BIO7 = Temperature Annual Range (BIO5-BIO6)
#BIO12 = Annual Precipitation
#BIO13 = Precipitation of Wettest Month
#BIO15 = Precipitation Seasonality (Coefficient of Variation)
Sample <- outlierdf[,1]
gdm_data <- cbind(Sample,num_out_df)
env_data <- subset(outlierdf, select = c(Sample, Latitude, Longitude, wc2.1_2.5m_bio_1, wc2.1_2.5m_bio_12, wc2.1_2.5m_bio_4, wc2.1_2.5m_bio_5, wc2.1_2.5m_bio_7, wc2.1_2.5m_bio_3, wc2.1_2.5m_bio_13, wc2.1_2.5m_bio_15))

gdmsub <- gdm_data[,100000:200000]

gdmsub <- cbind(Sample, gdmsub)

GDM <- formatsitepair(bioData = gdmsub, bioFormat = 1, dist="bray", abundance = T, siteColumn = "Sample", XColumn= "Latitude", YColumn = "Longitude", predData = env_data)
GDM_model <- gdm(GDM, geo=T)
plot(GDM_model)
summary(GDM_model1)

isplines_reduced <- as.data.frame(isplineExtract(GDM_model))


#GGplot 
Geographic_distance_total <- ggplot(isplines_reduced, aes(x=x.Geographic, y= y.Geographic)) + geom_line(aes(y = gdm_data, color="gdm_data", linewidth=2)) + xlab("Geographic Distance (degrees)") + ylab("f(Geographic Distance)") + theme_classic() + ylim(0,1)







gdm_varIMP<- gdm.varImp(GDM, geo=T) #calculates variants explained by each parameter. And gives summary statistic for the model
?gdm.varImp

## plotting the new plots for Bishop/Monterey
GDM_model1<- gdm(GDM_bish, geo=T)
plot(GDM_model1)
summary(GDM_model2)
GDM_model2 <- gdm(GDM_mont, geo=T)
plot(GDM_model2)
isplines_reduced1 <- as.data.frame(isplineExtract(GDM_model1))
isplines_reduced2 <- as.data.frame(isplineExtract(GDM_model2))

colors <- c("Bishop" = "salmon", "Monterey" = "skyblue")
Geographic_distance_Bishop <- ggplot(isplines_reduced1, aes(x=x.Geographic, y= y.Geographic)) + geom_line(aes(y = Bishop, color="Bishop", linewidth=2), size = 1.5) + xlab("Geographic Distance (degrees)") + ylab("f(Geographic Distance)") + theme_classic() + ylim(0,.1)

Precipitation_reduced_Bishop <- ggplot(isplines_reduced1, aes(x=x.wc2.1_2.5m_bio_15, y= y.wc2.1_2.5m_bio_15)) + geom_line(color="salmon", size=2) + xlab("Precipitation (coefficient of variance)") + ylab("f(Precipitation)") + theme_classic() + xlim(82,97) + ylim(0,.12)

Geographic_distance_Monterey <- ggplot(isplines_reduced2, aes(x=x.Geographic, y= y.Geographic)) + geom_line(color="skyblue", linewidth=2) + xlab("Geographic Distance (degrees)") + ylab("f(Geographic Distance)") + theme_classic() + xlim(0,1.4) + ylim(0,.015)

Temperature_Seasonality_Monterey <- ggplot(isplines_reduced2, aes(x=x.wc2.1_2.5m_bio_4, y= y.wc2.1_2.5m_bio_4)) + geom_line(color="skyblue", size=2) + xlab("Temperature (standard deviation x 100)") + ylab("f(Temperature)") + theme_classic() + xlim(200,350) + ylim(0,.1) 

library("ggpubr")
Figure1 <- ggarrange(Geographic_distance_Bishop, Precipitation_reduced_Bishop, Geographic_distance_Monterey, Temperature_Seasonality_Monterey, ncol = 2, nrow = 2)


```

```{r, GLM}

glm_data <- as.data.frame(num_out_df[, sample(1:ncol(num_out_df))])
env_data1 <- as.data.frame(cbind(env_data, glm_data))

Latitude = as.numeric(outlierdf$Latitude)
Longitude = as.numeric(outlierdf$Longitude)
colnames(env_data)[4] = "MAT"
colnames(env_data)[5] = "Annual Precipitation"
colnames(env_data)[6] = "Temp Seasonality"
colnames(env_data)[7] = "Max Temp"
colnames(env_data)[8] = "Annual Temp Range"
colnames(env_data)[9] = "Isothermality"
colnames(env_data)[10] = "Precipation of Wettest Month"
colnames(env_data)[11] = "Precipitation Seasonality"
env_data

snp_glm1 <- glm(glm_data[, 1] ~ Latitude, data=env_data1,family=binomial())
snp_glm1 <- glm(glm_data[, 1] ~ Latitude, data=env_data1,family=binomial())

write.table(env_data1, file = "env_data.csv", sep=",")
fwrite(glm_data, file = "glm_data_full.csv", sep=",", row.names = FALSE)


```

```{r, GWAS}
#https://adegenet.r-forge.r-project.org/files/Leuven2014/practical-GWAS.pdf
install.packages("glmnet", dep=TRUE)
install.packages('ade4')
install.packages('Matrix')
library(ade4)
library(adegenet)
library(stats)
library(glmnet)
library(Matrix)
?adegenet
snps <- as.matrix(num_out_df)
phen <- factor(outlierdf[,4])

#Create PCA
pca1 <- dudi.pca(snps, scale=FALSE)

#Plot PCA
s.label(pca1$li, sub="PCA - PC 1 and 2")
add.scatter.eig(pca1$eig,4,1,2, ratio=.3, posi="topleft")

#Plot the distances
D <- dist(pca1$li[,1:4])^2
clust <- hclust(D, method="complete")

temp <- as.data.frame(as.matrix(D))
temp <- t(as.matrix(D))
temp <- temp[,ncol(temp):1]


#Heat Map of samples
par(mar=c(1,5,5,1))
image(x=1:69, y=1:69, temp, col=rev(heat.colors(nlevels(as.factor(D)))),
xaxt="n", yaxt="n",
xlab="",ylab="")
axis(side=2, at=1:69, lab=rev(rownames(snps)), las=2, cex.axis=.46)
axis(side=3, at=1:69, lab=rownames(snps), las=2, cex.axis=.46)

#Dendrogram
plot(clust, main="Clustering (complete linkage) based on the first 4 PCs", cex=.4)

s.class(pca1$li, fac=phen, col=transp(c("royalblue","red")), cpoint=2,
sub="PCA - axes 1 and 2")
add.scatter.eig(pca1$eig,4,1,2, ratio=.24, posi="topleft")

#Calculate p-value
pval <- apply(snps, 2, function(e)
  fisher.test(table(factor(e, levels=c(0,1)), phen))$p.value)
#Corrected p-value
pval.corrected <- p.adjust(pval, method="fdr")
length(which(pval.corrected < 0.05))

snps.selected.univariate <- which(pval.corrected < 0.05)
n.snps.selected.univariate <- length(snps.selected.univariate)

snps.selected.univariate
str(snps.selected.univariate)

#Manhattan Plot using Bonferroni Correction
log.pval <- -log10(pval)
jitter.pval <- jitter(pval, amount=0.2)
set.seed(1)
log.pval <- jitter(log.pval, amount=0.2)
plot(log.pval,
col = c("gray"),
pch = 19,
cex = 0.5,
main="Manhattan plot: Fisher's exact test with Bonferroni correction",
xlab="SNP loci", ylab="Uncorrected -log10(p-value)",
cex.main=1)
bonferroni <- -log10(0.05 / ncol(snps))
abline(h=bonferroni, col = "black")

#Manhattan Plot using False Discovery Rate (FDR)
log.pval <- -log10(pval.corrected)
set.seed(1)
log.pval <- jitter(log.pval, amount=0.2)
plot(log.pval,
col = c("gray"),
pch = 19,
cex = 0.5,
main="Manhattan plot: Fisher's exact test with FDR",
xlab="SNPs", ylab="Corrected -log10(p-value)")


#Manhattan qqman package
pval_df <- as.data.frame(jitter.pval)
pval_gwas <- as.numeric(pval_df$jitter.pval)
snp_pos <- rownames(pval_df)

GWAS <- cbind(adapt_POS, adapt_CHROM, adapt_BP, jitter.pval) #pval_gwas
colnames(GWAS)[1] = "POS"
colnames(GWAS)[2] = "CHROM"
colnames(GWAS)[3] = "BP"
colnames(GWAS)[4] = "P"
#adapt_gwas <- na.omit(gwas)
GWAS <- GWAS[gwas$P >= 0, ]
GWAS <- na.omit(GWAS)
out_pcadapt <- sub("^", "sp", outliers_pcadapt)
manhattan(GWAS, chr="CHROM", bp="BP", snp="POS", p="P", main = "Manhattan Plot", ylim = c(0,6), cex = 0.8, suggestiveline = T, genomewideline = T, highlight = pcadapt_outflank_snps) #NEED TO REPLACE OUTLIER_PCADAPT WITH SCAFFOLD_SNP POSITIONS




```

```{r, OutFLANK}


library(devtools)
library(qvalue)
library(vcfR)
library(OutFLANK)
library(qqman)
#http://rstudio-pubs-static.s3.amazonaws.com/305384_9aee1c1046394fb9bd8e449453d72847.html
#https://popgen.nescent.org/2016-01-26-SNP-selection.html
gwasResults
num_out_df
num_matrix <- data.matrix(num_out_df, rownames.force = NA)
colnames(num_matrix) <- NULL

outlierdf
ind <- outlierdf[,4]
lociname <- as.character(colnames(num_out_df))
FstDataFrame <- MakeDiploidFSTMat(num_matrix, lociname, ind)


#######This is all visualizations of data########
#plot(FstDataFrame$FST, FstDataFrame$FSTNoCorr, xlim = c(-0.01,0.3), 
    # ylim = c(-0.01, 0.3), pch = 20)
#abline(0, 1) 
#hist(FstDataFrame$FSTNoCorr) 

#hist(FstDataFrame$FSTNoCorr) 
#plot(FstDataFrame$LocusName, FstDataFrame$FST, ylim = c(0.01, 0.8), xlim = c(0, 1), pch=20, col="grey")
#hist(FstDataFrame$FSTNoCorr, breaks=seq(0,0.6, by=0.001))
#hist(FstDataFrame$FSTNoCorr[FstDataFrame$He>0.05], breaks=seq(0,0.6, by=0.001))
#hist(FstDataFrame$FSTNoCorr[FstDataFrame$He>0.1], breaks=seq(0,0.6, by=0.001))


#OF <- OutFLANK(FstDataFrame, NumberOfSamples=69, qthreshold = 0.01, 
               #RightTrimFraction = 0.05)
#OutFLANKResultsPlotter(OF, withOutliers = TRUE, NoCorr = TRUE, Hmin = 0.1, 
                      # binwidth = 0.005, Zoom = FALSE, RightZoomFraction = 0.05, 
                      # titletext = NULL)
#########
#outliers_OF <- OF$results$LocusName[OF$results$OutlierFlag == TRUE]
#length(outliers_OF)
#########

outlier <- OutFLANK(FstDataFrame,NumberOfSamples = 2, 
                     RightTrimFraction = 0.06, LeftTrimFraction = 0.35,
                     qthreshold = 0.05, Hmin = 0.1)
OutFLANKResultsPlotter(outlier, withOutliers = TRUE, NoCorr = TRUE, Hmin = 0.1, 
                       binwidth = 0.005, Zoom = FALSE, RightZoomFraction = 0.05, 
                       titletext = NULL)
outlier_OF <- outlier$results$LocusName[outlier$results$OutlierFlag == TRUE]
print(outlier_OF)
length(outlier_OF)


outflank_results <- as.data.frame(outlier$results)
#write.csv(outflank_results, file = "OUTFLANK_results.csv")
nrow(FstDataFrame)
nrow(outlier$results)
sum(outlier$results$qvalues<0.05, na.rm=TRUE)

plot(outlier$results$He, outlier$results$FST, pch=20, col="grey")
  points(outlier$results$He[outlier$results$qvalues<0.05], y = outlier$results$FST[outlier$results$qvalues<0.05], pch=21, col="blue")

outlier    
top_candidates <- outlier$results$qvalues<0.05 & outlier$results$He>0.1  
topcan <- outlier$results[top_candidates,]
outlier_loci <- topcan[order(topcan$LocusName),]

outflank_snp_pos <- rownames(topcan)
#make the outlier_snp column based off the rownames
outflank_snps <- rownames(topcan)
sp_outflank_snps <- sub("^", "sp", row.names(topcan))
#Making outlier loci df compatible with gwas 
split_outlier_loci <- as.data.frame(str_split_fixed(outlier_loci$LocusName, "_", 3))
out_chrom <- as.numeric(split_outlier_loci[,2])
out_pos <- split_outlier_loci[,3]
sp_pos <- sub("^", "sp", out_pos) #This is highlighting the snp position, not the rowname
out_loci_df <- as.data.frame(cbind(sp_pos, out_chrom))
colnames(out_loci_df)[1] <- "POS"
colnames(out_loci_df)[2] <- "CHROM"
out_loci_df
str(out_loci_df)
row_topcan <- rownames(outlier$results[top_candidates,])


#Manhattan Plot
snp_num <- outlier$results$LocusName
out_loci <- row.names(outlier_loci)
library(stringr)
split <- as.data.frame(str_split_fixed(outlier$results$LocusName, "_", 3))
CHROM <- as.numeric(split[,2])
POS <- split[,3]
num_out_df
BP <- as.data.frame(seq(nrow(outlier$results)))
P <- -log(outlier$results$qvalues)     
snp_position <- sub("^", "sp", BP$`seq(nrow(outlier$results))`)

gwas <- cbind(snp_position, CHROM, BP, P)
colnames(gwas)[1] = "POS"
colnames(gwas)[2] = "CHROM"
colnames(gwas)[3] = "BP"
colnames(gwas)[4] = "qvalues"
gwas <- na.omit(gwas)

gwas <- gwas[gwas$P >= 0, ]
out_loci <- sub("^", "sp", row.names(outlier_loci))
manhattan(gwas, chr="CHROM", bp="BP", snp="POS", p="qvalues", main = "Manhattan Plot", ylim = c(0,7), cex = 0.8, suggestiveline = T, genomewideline = T, logp=FALSE, ylab="qvalues", highlight = pcadapt_outflank_snps)



```

```{r, PCAdapt}
#https://popgen.nescent.org/2016-01-26-SNP-selection.html
library(pcadapt)

pca_genotype <- read.pcadapt(t(num_out_df))
str(pca_genotype)
K <- 25
x <- pcadapt(pca_genotype, K = 4, method = c("mahalanobis"), min.maf = 0.05, ploidy =2, LD.clumping = list(size = 200, thr = 0.1))
summary(x)
plot(x, option = "qqplot", threshold = 0.1)
plot(x, option = "stat.distribution") # Distribution of Mahalanobis distances.
str(x)
padj <- p.adjust(x$pvalues,method="bonferroni")
alpha <- 0.05
outliers_pcadapt <- which(padj < alpha)
length(outliers_pcadapt)

AdaptFlank_outliers <- as.data.frame(outliers_pcadapt[outliers_pcadapt %in% outflank_snp_pos])
##############################GLM##############################
temp <- outlierdf[ ,7]
outlier_loci$LocusName[1] #From outflank
outliers_pcadapt #this is from PCAdapt
out_loci1 <- as.data.frame(as.numeric(out_loci))
loc1 <- num_out_df[ ,pcadapt_snps[16]]
loc1temp <- data.frame(loc1, temp)
mod <- glm(cbind(loc1, 2 - loc1) ~ temp, family = binomial)

ggplot(loc1temp, aes(x = factor(loc1), y = temp)) + 
 geom_boxplot() + 
 xlab("Major allele count") +
 ylab("Mean Annual Temperature")

```

```{r, qqman}

library(qqman)
str(gwasResults)
tail(gwasResults)
snpsOfInterest
adapt_gwas


```

```{r, Manhattan Plots for PCADAPT}
#Manhattan Plot
library(stringr)
split <- as.data.frame(str_split_fixed(snp_pos, "_", 3)) #snp_pos is generated from GWAS section
adapt_CHROM <- as.numeric(split[,2])
adapt_BP <- as.data.frame(seq(ncol(num_out_df)))
adapt_POS <- as.factor(sub("^", "sp", adapt_BP$`seq(ncol(num_out_df))`))
adapt_P <- x$pvalues
num_out_df

#the snp positions of the PCAdapt outliers
pcadapt_snps <- outlier$results[outliers_pcadapt, 1]


write.table(pcadapt_snps, file = "Outloci_pos_pcadapt_0.05.csv", append =FALSE, sep = " ", dec = ".", row.names = TRUE, col.names = TRUE)

length(unique(pcadapt_snps))

outflank_snps <- outlier_loci$LocusName

outflank_snp_pos1 <- as.numeric(outflank_snp_pos)
AdaptFlank_outliers <- as.data.frame(pcadapt_snps[pcadapt_snps %in% outflank_snps])
pcadapt_outflank_combined <- as.data.frame(outliers_pcadapt[outliers_pcadapt %in% outflank_snp_pos1])
pcadapt_outflank_snps <- sub("^", "sp", pcadapt_outflank_combined$`outliers_pcadapt[outliers_pcadapt %in% outflank_snp_pos1]`)



adapt_gwas <- cbind(adapt_POS, adapt_CHROM, adapt_BP, adapt_P) #pval_gwas
colnames(adapt_gwas)[1] = "POS"
colnames(adapt_gwas)[2] = "CHROM"
colnames(adapt_gwas)[3] = "BP"
colnames(adapt_gwas)[4] = "P"
#adapt_gwas <- na.omit(gwas)
#gwas <- gwas[gwas$P >= 0, ]
out_pcadapt <- sub("^", "sp", outliers_pcadapt)
manhattan(adapt_gwas, chr="CHROM", bp="BP", snp="POS", p="P", main = "Manhattan Plot", ylim = c(0,12), cex = 0.8, suggestiveline = T, genomewideline = T, highlight = pcadapt_outflank_snps)

```

```{r, Bayescan}
outlierdf
bayescan_inputfile <- as.data.frame(cbind(outlierdf[3], outlierdf[5], num_out_df))
colnames(bayescan_inputfile) <- NULL
rownames(bayescan_inputfile) <- NULL

bayescan_test <- as.data.frame(cbind(outlierdf[3], outlierdf[5], num_out_df[, sample(1:ncol(num_out_df), 10)]))
fwrite(bayescan_test, file = "/Users/jyeam/Documents/Projects/2022_Population_Genomics/R/2023_Population_Genomics_SP_fc27/bayescan_test.txt", append =FALSE, sep = " ", quote = TRUE, row.names = FALSE, col.names = FALSE)

fwrite(bayescan_inputfile, file = "/Users/jyeam/Documents/Projects/2022_Population_Genomics/R/2023_Population_Genomics_SP_fc27/bayescan_inputfile.txt", append =FALSE, sep = " ", quote = TRUE, row.names = FALSE, col.names = FALSE)

install.packages("bayescan.R")

```


``` {r, Maps}

#https://rstudio-pubs-static.s3.amazonaws.com/676340_b6f5780eb85046ad88a6f59732068631.html
#devtools::install_github("dkahle/ggmap")
#install.packages('mapdata')

library(ggplot2)
library(ggmap)
library(maps)
library(mapdata)
library(stringr)
library(dplyr)
library(tidyverse)
library(raster)

??mapdata

usa <- map_data("usa")
states <- map_data("state")
ca_df <- subset(states, region == "california")
counties <- map_data("county")
ca_county <- subset(counties, region == "california")
SampleCoordinates <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/Map_Count.csv", stringsAsFactors = FALSE)
ca_base <- ggplot(data = ca_df, mapping = aes(x = long, y = lat, 
                group = group)) + 
                coord_fixed(1.3) + 
                geom_polygon(color = "black", fill = "gray")

ca_base + theme_nothing() + 
  geom_polygon(data = ca_county, fill = NA, color = "white") +
  geom_polygon(color = "black", fill = "gray") + 
  geom_point(data = SampleCoordinates, aes(x=Longitude, y=Latitude, size=Count), inherit.aes =FALSE)



ggplot(data = ca_df, mapping = aes(x = long, y = lat, group = group)) + 
                coord_fixed(1.3) + 
                geom_polygon(color = "black", fill = "gray") +
geom_point(data = SampleCoordinates, colour = "red", size=Count))

ll_means <- sapply(sisquoc[2:3], mean)
sq_map2 <- get_map(location = ll_means,  maptype = "satellite", source = "google", zoom = 15)

###########################################

US <- getData("GADM",country="USA",level=1)
states    <- c('California')
california <- US[US$NAME_1 %in% states,]

PNW <- US[US$NAME_1 %in% states,]
plot(california, axes=TRUE)
ggplot(california) + geom_polygon(data=california, aes(x=long,y=lat,group=group),
  fill="cadetblue", color="grey") + coord_equal()
US$NAME_1
download.file("https://github.com/mhweber/AWRA_GIS_R_Workshop/blob/gh-pages/files/SRTM_OR.RData?raw=true",
              "SRTM_OR.RData",
              method="auto",
              mode="wb")

load("./SRTM_OR.RData")
srtm <- getData('SRTM', lon=-122, lat=41)
srtm1 <- getData('SRTM', lon=-121, lat=40)
srtm2 <- getData('SRTM', lon=-120, lat=37.3)
srtm3 <- getData('SRTM', lon=-118.5, lat=35)
srtm4 <- getData('SRTM', lon=-117, lat=34)
srtm5 <- getData('SRTM', lon=-115, lat=34)
srtm6 <- getData('SRTM', lon=-120, lat=34)
srtm_all <- mosaic(srtm, srtm1, srtm2, srtm3, srtm4, srtm5, srtm6, fun = mean)

plot(srtm_all)
plot(california, add=TRUE)

srtm_crop_cal <- crop(srtm_all, california)
srtm_mask_cal <- mask(srtm_crop_cal, california)
plot(srtm_mask_cal, main="Elevation (m) in California")
plot(california, add=TRUE)


```

```{r, Map with georeference data}
#https://www.youtube.com/watch?v=Of_M4kcE9yM&ab_channel=StatisticsGuideswithDrPaulChristiansen


library(rstudioapi)
library(ggmap)
register_google(key = "AIzaSyCK_g3QY-OOmNogSUnWbga1DQNKdadtFO0")
SampleCoordinates <- read.csv("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Files/Map_Count.csv", stringsAsFactors = FALSE)

cal_box <- make_bbox(lat = Latitude, lon = Longitude, data = SampleCoordinates)

cal_big <- get_map(location = cal_box, source = "google", maptype = "terrain")
get_map()
cal_zoom <- get_map(location = c(lon = -121, lat = 38), zoom = 7, source = "google", maptype = "terrain")
ggmap(cal_zoom) + 
  geom_point(data = SampleCoordinates, mapping = aes(x = Longitude, y = Latitude, color = "blue", size = Count))

get_map()



ca_base + theme_nothing() + 
  geom_polygon(data = ca_county, fill = NA, color = "white") +
  geom_polygon(color = "black", fill = "gray") + 
  geom_point(data = SampleCoordinates, aes(x=Longitude, y=Latitude, size=Count), inherit.aes =FALSE)
```



```{r, GO terms}

#gene information for all snps from vcf
snp.gene <- read.delim("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/SP_snpgenes.gff", header=F, comment.char="#")
#Gene Ontology text file
GO.df <- as.data.frame(read.table("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/Suipun1/Suipun1_GeneCatalog_proteins_20200424_GO.tab", header=F, sep="\t", fill = TRUE, quote = ""))
KOG.df <- as.data.frame(read.table("/Users/jyeam/Documents/Projects/2022_Population_Genomics/Downloads/Suipun1/Suipun1_GeneCatalog_proteins_20200424_KOG.tab.gz", header=F, sep="\t", fill = TRUE, quote = ""))



#removing some columns and seperating INFO column
gene_list <- data.frame(snp.gene$V1, snp.gene$V2, snp.gene$V3, snp.gene$V4, snp.gene$V8, do.call(rbind, strsplit(snp.gene$V9, split = ";", fixed = TRUE)))

#remove unnecessary columns
gene_list1 <- subset(gene_list, select = -c(X1, X3, snp.gene.V8))

#rename columns
colnames(gene_list1)[1] = "Chrom"
colnames(gene_list1)[2] = "Database"
colnames(gene_list1)[3] = "Type"
colnames(gene_list1)[4] = "Position"
colnames(gene_list1)[5] = "Protein_ID"
colnames(gene_list1)[6] = "Product_Name"

gene_list1$Product_Name <- gsub("product_name","", gene_list1$Product_Name)
gene_list1$Protein_ID <- gsub("proteinId", "", gene_list1$Protein_ID)
gene_list1
length(unique(gene_list1$Type))
#remove rows that are in the exon
gene_list2 <-gene_list1[!(gene_list1$Type=="exon"), ]
gene_list2

gene_list2$Protein_ID <- as.numeric(gene_list2$Protein_ID)
################################################
colnames(GO.df)[1] = "proteinId"
colnames(GO.df)[2] = "gotermId"
colnames(GO.df)[3] = "goName"
colnames(GO.df)[4] = "gotermType"
colnames(GO.df)[5] = "goAcc"
################################################


#Pull goterms that are in gene_list2 dataframe
GO_term_snp_list <- GO.df %>% 
  filter(GO.df$proteinId %in% gene_list2$Protein_ID)


length(unique(GO_term_snp_list$proteinId))
length(unique(gene_list2$Protein_ID))

pcadapt_snps
pcadapt_df <- as.data.frame(pcadapt_snps)

pcadapt_df1 <- pcadapt_df %>% 
  separate(pcadapt_snps, into = c("Chromosome", "Position"), sep="_(?=[^_]+$)")

pcadapt_df1$Position <- as.integer(pcadapt_df1$Position)
pcadapt_df1


pcadapt_loci <- gene_list2[gene_list2$Chrom %in% pcadapt_df1$Chromosome & gene_list2$Position %in% pcadapt_df1$Position, ]



PCAdapt_GO_terms <- GO.df %>% 
  filter(GO.df$proteinId %in% pcadapt_loci$Protein_ID)
length(unique(gene_list$Protein_ID))



#PCAdapt outlier snps mapped to gene identify
Outliers_pcadapt_gff <- read.delim("/Users/jyeam/Documents/Projects/2022_Population_Genomics/R/2023_Population_Genomics_SP_fc27/Outloci_pcadapt_0.05.gff", header=F, comment.char="#")
#removing some columns and seperating INFO column for PCAdapt GFF
PCAdapt_gene_list <- data.frame(Outliers_pcadapt_gff$V1, Outliers_pcadapt_gff$V2, Outliers_pcadapt_gff$V3, Outliers_pcadapt_gff$V4, Outliers_pcadapt_gff$V8, do.call(rbind, strsplit(Outliers_pcadapt_gff$V9, split = ";", fixed = TRUE)))

#remove unnecessary columns
pcadapt_gene_list1 <- subset(PCAdapt_gene_list, select = -c(X1, X3, Outliers_pcadapt_gff.V8))

#rename columns
colnames(pcadapt_gene_list1)[1] = "Chrom"
colnames(pcadapt_gene_list1)[2] = "Database"
colnames(pcadapt_gene_list1)[3] = "Type"
colnames(pcadapt_gene_list1)[4] = "Position"
colnames(pcadapt_gene_list1)[5] = "Protein_ID"
colnames(pcadapt_gene_list1)[6] = "Product_Name"

pcadapt_gene_list1$Product_Name <- gsub("product_name","", pcadapt_gene_list1$Product_Name)
pcadapt_gene_list1$Protein_ID <- gsub("proteinId", "", pcadapt_gene_list1$Protein_ID)
pcadapt_gene_list1
length(unique(pcadapt_gene_list1$Type))
#remove rows that are in the exon
pcadapt_gene_list2 <-pcadapt_gene_list1[!(pcadapt_gene_list1$Type=="exon"), ]
pcadapt_gene_list3 <-pcadapt_gene_list2[!(pcadapt_gene_list2$Type=="stop_codon"), ]
pcadapt_gene_list3

pcadapt_gene_list3$Protein_ID <- as.numeric(pcadapt_gene_list3$Protein_ID)

PCAdapt_GO <- GO.df %>% 
  filter(GO.df$proteinId %in% pcadapt_gene_list3$Protein_ID)
View(PCAdapt_GO)
length(unique(PCAdapt_GO$proteinId))


```



